### ISSUE TITLE
Suspected information leakage in training & val

### ISSUE THREAD

**Issue by ChangyaoTian** [buxiangzhiren/DDCap#1]
Hello, I am a little confused about your implementation [here](https://github.com/buxiangzhiren/DDCap/blob/main/train.py#L359).

```python
def forward(self, tokens: torch.Tensor, mask_tokens: torch.Tensor, prefix: torch.Tensor,
                mask: Optional[torch.Tensor] = None, t = None,
                labels: Optional[torch.Tensor] = None):
      self.clip_model.eval()
      embedding_text = self.gpt.transformer.wte(tokens)
      batch_size = embedding_text.size()[0]
      seq_len = embedding_text.size()[1]
      bos_token_embedding = self.bos_embedding.unsqueeze(0).unsqueeze(0).repeat_interleave(repeats=batch_size, dim=0)
      bos_token_embedding = bos_token_embedding.repeat_interleave(repeats=seq_len, dim=1)
      mask_tokens = mask_tokens.unsqueeze(dim=2).repeat_interleave(repeats=self.gpt_embedding_size, dim=2)
      embedding_text = torch.where(mask_tokens == 50257, bos_token_embedding, embedding_text)
      with torch.no_grad():
          prefix, len_cls = self.image_encode(prefix)
      prefix_projections = self.clip_project(prefix)
      if self.training:
          empty_idx = int(self.if_drop_rate * batch_size)
          for i_image in range(empty_idx):
              prefix_projections[i_image,:,:] = self.pad_embedding
      if labels is not None:
          dummy_token = self.get_dummy_token(tokens.shape[0], tokens.device)
          labels = torch.cat((dummy_token, tokens), dim=1)
      out = self.gpt(t=t, inputs_embeds=embedding_text, labels=labels, attention_mask=mask,
                     encoder_hidden_states=prefix_projections)
      return out, self.len_head(len_cls)
```

It seems that the gpt model always takes the embeddings of `tokens` rather than `mask_tokens` as input during both train and val stage. However, if I am not mistaken, these `tokens` _directly_ come from **the gt caption**, while the `mask_tokens` instead are sampled from the diffusion process. This may induce serious information leakage and thus make the performance result in your paper less convincing.

Could you explain more about this ? Thanks a lot!

**Comment by buxiangzhiren** [comment]
The line of ''embedding_text = torch.where(mask_tokens == 50257, bos_token_embedding, embedding_text)" replace the embeddings of tokens from mask_tokens.

**Comment by buxiangzhiren** [comment]
In the process of validation, the embeddings of tokens which come from the gt caption are just used to calculate loss. The text generation used to compute CIDEr is the function "generate2_adpt_if", which starts at mask_tokens insted of gt. 
![image](https://user-images.githubusercontent.com/67734862/203689339-691d8d92-2e4e-4fad-ad73-59a6cb2c0d35.png)

**Comment by ChangyaoTian** [comment]
Thanks for your reply. Sorry I was wrong about the information leakage thing in validation. But I am still confused about your whole diffusion process.

```python
embedding_text = torch.where(mask_tokens == 50257, bos_token_embedding, embedding_text)
```
This line means only those tokens whose id is 50257 (\<mask\> or \<bos\>?) are replaced by bos_token_embedding, but all other tokens (no matter corrupted or not), remain the same as gts. So, the input tokens are just the combination of \<mask\> and gt tokens. Such implementation seems quite different from the VQ-diffusion process, whose input also contains corrupted tokens.

In the process of validation, although the gt tokens are not used, the input are still the combination of \<mask\> and predicted tokens, without the q_posterior process used in standard VQ-diffusion process, which will resample the tokens from your predicted ones. 

To sum up, it seems that you design a dynamic masking strategy for training and inference instead of using the diffusion process. Is that true?

**Comment by buxiangzhiren** [comment]
1. The <mask> token is the corrupted tokens
2.  The q_posterior process used in standard VQ-diffusion process is not work well, which was explained in the paper. We also improved the implementation of q_posterior process in the function "generate_diffsuion" in misc.py. You could use it instead of "generate2_adpt_if" to generate text and obtain the score of CIDEr.
![image](https://user-images.githubusercontent.com/67734862/203693235-5935cf10-3cd6-4d8f-9732-dbca4e6798cb.png)

**Comment by buxiangzhiren** [comment]
I am so sorry for my misunderstanding of your problems. I added a new line which can help you understand our code. Since the probability of beta in our noise schedule is very small and according to the conclusion through VQ-diffusion, the diffusion process with or without beta which just cause slight influence. So we don't use it in our code.
![image](https://user-images.githubusercontent.com/67734862/203704745-85c6a26e-9aae-4463-96da-4e8851388a24.png)

**Comment by ChangyaoTian** [comment]
Fine, now I understand what you have done here:
1. Regarding the diffusion process, since the beta prob is relatively tiny, you directly ignore it and only consider the masking prob gamma.

2. During the posterior sampling process, you replace the original q_posterior function with your best-first inference strategy. 

However, the fact is that the two designs (i.e., q_sampling & q_posterior) above are exactly what a standard diffusion process needs to have. So these changes may not be consistent with some points you claim in your paper, such as:

![image](https://user-images.githubusercontent.com/97594817/203883204-a83e5a29-5074-4968-bff4-d2c53c0ab140.png)

All these comments are just a kind reminder. Thanks for your reply.

**Comment by buxiangzhiren** [comment]
Thanking for your reminder. There are still two important things.
1. We indeed obtain the x_t via the process of q_sampling which is the standard diffusion process.
2. In inference, the reason why we replace the original q_posterior function with our best-first inference strategy is that it is not workable. The point is that our best-first inference strategy maintains the same noise ratio as q_posterior in the process of denoising, _i.e._, the q_posterior function. The idea of gradual de-noising as the core of the diffusion processs has not changed
3. These changes are described in Section 3.2 and consistent with points claimed in our paper. The picture you showed is just the original inference process in Section 3.1 instead of ours. In addition, we also improved the code of original q_posterior function to conduct inference.

**Comment by buxiangzhiren** [comment]
Anyway, thank you for your valuable opinions.
